src/lib/services/OutputProcessor.ts

```diff
@@ -64,27 +85,47 @@ export class OutputProcessor extends AsyncCreatable {
 	}
 
 	// TODO: consider moving all message creation logic to a separate place and making this method private
-	public emitEvents(outEvents: RuleEvent[]): boolean {
+	public async emitEvents(outEvents: RuleEvent[]): Promise<boolean> {
 		this.logger.trace('About to order and emit');
 		// If list is empty, we can just be done now.
 		if (outEvents.length == 0) {
 			return false;
 		}
 
 		// Iterate over all of the events and throw them as appropriate.
+		const telemetryEvents: Promise<void>[] = [];
 		outEvents.forEach((event) => {
 			this.logEvent(event);
-			let eventType = '';
 			if (event.handler === 'UX_SPINNER') {
-				eventType = EVENTS.UPDATE_SPINNER.toString();
+				this.emitUxEvent(EVENTS.UPDATE_SPINNER.toString(), event.messageKey, event.args);
 			} else if (event.handler === 'UX' || (event.handler === 'INTERNAL' && event.type === 'ERROR')) {
-				eventType = `${event.type.toLowerCase()}-${event.verbose ? 'verbose' : 'always'}`;
+				const eventType = `${event.type.toLowerCase()}-${event.verbose ? 'verbose' : 'always'}`;
+				this.emitUxEvent(eventType, event.messageKey, event.args);
+			} else if (event.handler === 'INTERNAL' && event.type === 'TELEMETRY') {
+				telemetryEvents.push(this.emitTelemetryEvent(event.args));
 			}
-			this.emitUxEvent(eventType, event.messageKey, event.args);
 		});
+		await Promise.all(telemetryEvents);
 		return true;
 	}
 
+	private async emitTelemetryEvent(args: string[]): Promise<void> {
```

Comment:
This method being `async` required several other methods to be `async`, and that cascaded upwards through several other files.